include/seqan3/alignment/matrix/detail/aligned_sequence_builder.hpp

```diff
@@ -0,0 +1,250 @@
+ * expressions must evaluate to `true` in order use the sequences with the given aligned sequence types:
+ * ```cpp
+ * std::Constructible<typename traits_t::fst_aligned_t, decltype(std::declval<fst_rng_t>() | view::slice(0, 1))>
+ * std::Constructible<typename traits_t::sec_aligned_t, decltype(std::declval<sec_rng_t>() | view::slice(0, 1))>
+ * ```
+ */
+template <std::ranges::ViewableRange fst_rng_t,
+          std::ranges::ViewableRange sec_rng_t,
+          typename traits_t = aligned_sequence_builder_default_traits<fst_rng_t, sec_rng_t>>
+class aligned_sequence_builder
+{
+private:
+
+    static_assert(std::Constructible<typename traits_t::fst_aligned_t,
+                                     decltype(std::declval<fst_rng_t>() | view::slice(0, 1))>,
+                  "The given sequence type cannot be transformed to a AlignedSequence. It needs to be a view::slice "
+                  "over the original type.");
+
+    static_assert(std::Constructible<typename traits_t::sec_aligned_t,
+                                     decltype(std::declval<sec_rng_t>() | view::slice(0, 1))>,
+                  "The given sequence type cannot be transformed to a AlignedSequence. It needs to be a view::slice "
+                  "over the original type.");
+
+    using fst_aligned_t = typename traits_t::fst_aligned_t; //!< The aligned sequence type for the first range.
+    using sec_aligned_t = typename traits_t::sec_aligned_t; //!< The aligned sequence type for the second range.
+
+public:
+
+    //!\brief The result type when building the aligned sequences.
+    struct result_type
+    {
+        matrix_coordinate begin; //!< The coordinate where the trimmed first and second sequence begin.
+        matrix_coordinate end; //!< The coordinate where the trimmed first and second sequence end.
+        std::pair<fst_aligned_t, sec_aligned_t> alignment; //!< The alignment over the trimmed sequences.
+    };
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr aligned_sequence_builder() = default; //!< Defaulted.
+    constexpr aligned_sequence_builder(aligned_sequence_builder const &) = default; //!< Defaulted.
+    constexpr aligned_sequence_builder(aligned_sequence_builder &&) = default; //!< Defaulted.
+    constexpr aligned_sequence_builder & operator=(aligned_sequence_builder const &) = default; //!< Defaulted.
+    constexpr aligned_sequence_builder & operator=(aligned_sequence_builder &&) = default; //!< Defaulted.
+    ~aligned_sequence_builder() = default; //!< Defaulted.
+
+    /*\brief Construction from the underlying sequences
+     * \param[in] fst_rng The first range to build the aligned sequence for.
+     * \param[in] sec_rng The second range to build the aligned sequence for.
+     */
+    constexpr aligned_sequence_builder(fst_rng_t fst_rng, sec_rng_t sec_rng) :
+        fst_rng{view::all(std::forward<fst_rng_t>(fst_rng))},
```

Comment:
why do you store views to the ranges, instead of passing them to `build_from`?

Reply:
So I can build over the same sequences different paths, e.g. when searching for sub optimal alignments or reporting all paths with the same score.

Reply:
but whats better in this 
```cpp
sequence_builder sb{seq1, seq2};
for (auto & trace : all_traces_of_interest)
    sb.build_from(trace);
```
compared to this
```cpp
for (auto & trace : all_traces_of_interest)
    build_alignment(seq1, seq2, trace);
```

The only thing I can see is that you would need to pass one variable (sequence_builder) to subsequent functions, instead of two (the two sequences.)?

I mean If you say all of this fits much better in your alignment design I'm fine with this, but I try to start keeping track of general design decisions to learn and apply them in the future. And I would have designed this as a free detail function and I am not convinced yet about the advantages :grin: help me out.